python/utils/bootstrap_utilities.py

```diff
@@ -391,3 +392,160 @@ def handle_error(data, proxy=None):
+                else:
+                    current_dict[key] = value
+    return config
+
+
+def execute_pre_initialization_hook(data):
+    """
+    Executes the 'hook_pre_initialization' hook.
+
+    This function dynamically loads and executes the 'hook_pre_initialization' hook,
+    which is responsible for performing early initialization tasks before the engine
+    is fully initialized. These tasks may include importing necessary libraries,
+    setting environment variables, or preparing the environment to ensure compatibility
+    with PySide6 and other dependencies.
+
+    If a custom hook is not defined in the advanced configuration (`tk-desktop.yml`),
+    the function falls back to dynamically loading the default hook located in the
+    'hooks' directory of the project.
+
+    The function follows these steps:
+    1. Reads the `info.yml` file to retrieve the default hook name for pre-initialization.
+    2. Reads the `tk-desktop.yml` file from the advanced configuration to check if a
+       custom hook is defined for the project.
+    3. Resolves the path to the custom hook if defined, or falls back to the default hook.
+    4. Dynamically loads the hook module and executes its `PreInitialization.execute` method.
+
+    :param data: A dictionary containing configuration and runtime details, including:
+                 - `config_path`: The root path of the advanced configuration.
+                 - `proxy_data`: Proxy connection details for logging.
+                 - `rpc_lib_path`: Path to the RPC library for communication.
+    """
+
+    try:
+        # Step 1: Read info.yml to get the default hook name
+        info_yml_path = os.path.join(os.path.dirname(__file__), "..", "..", "info.yml")
+        info = parse_yml_file(info_yml_path)
+        default_hook_name = (
+            info.get("configuration", {})
+            .get("hook_pre_initialization", {})
+            .get("default_value", "")
+        )
+
+        # Step 2: Read tk-desktop.yml to check for a custom hook
+        tk_desktop_yml_path = os.path.join(
+            data.get("config_path"),
+            "config",
+            "env",
+            "includes",
+            "settings",
+            "tk-desktop.yml",
+        )
+        tk_desktop_config = parse_yml_file(tk_desktop_yml_path)
+        hook_path = (
+            tk_desktop_config.get("settings.tk-desktop.project", {})
+            .get("hooks", {})
+            .get(default_hook_name)
+        )
+
+        # Step 3: Fallback to the default hook if no custom hook is defined
+        if hook_path and not os.path.exists(hook_path):
+            message = (
+                f"Custom pre-initialization hook specified is set to '{hook_path}',"
+                " but the file does not exist. "
+                "Falling back to the default pre-initialization hook. pipiree"
+            )
+            logger_via_proxy(data, message)
+            hook_path = None
+
```

Comment:
Looks good but I still don't understand why we can't call methods of tk-core code here instead of duplicating the code here?

I know that's because PySide6 (QtImporter) is called early on but I reviewed the code about Qt imports recently and there is definitely a big part of core that is supposed to be used **without** Qt/PySide